src/audio/kpb.c

```diff
@@ -366,61 +369,58 @@ static int kpb_reset(struct comp_dev *dev)
 static int kpb_copy(struct comp_dev *dev)
 {
 	int ret = 0;
-	int update_buffers = 0;
 	struct comp_data *kpb = comp_get_drvdata(dev);
 	struct comp_buffer *source;
 	struct comp_buffer *sink;
+	size_t copy_bytes = 0;
 
 	tracev_kpb("kpb_copy()");
 
-	/* get source and sink buffers */
+	/* Get source and sink buffers */
 	source = list_first_item(&dev->bsource_list, struct comp_buffer,
 				 sink_list);
-	sink = kpb->rt_sink;
-
-	/* process source data */
-	/* check if there are valid pointers */
-	if (source && sink) {
-		if (!source->r_ptr || !sink->w_ptr) {
-			return -EINVAL;
-		} else if (sink->free < kpb->sink_period_bytes) {
-			trace_kpb_error("kpb_copy() error: "
-				   "sink component buffer"
-				   " has not enough free bytes for copy");
-			comp_overrun(dev, sink, kpb->sink_period_bytes, 0);
-			/* xrun */
-			return -EIO;
-		} else if (source->avail < kpb->source_period_bytes) {
-			/* xrun */
-			trace_kpb_error("kpb_copy() error: "
-					   "source component buffer"
-					   " has not enough data available");
-			comp_underrun(dev, source, kpb->source_period_bytes,
-				      0);
-			return -EIO;
-		} else {
-			/* sink and source are both ready and have space */
-			/* TODO: copy sink or source period data here? */
-			memcpy(sink->w_ptr, source->r_ptr,
-			       kpb->sink_period_bytes);
-			/* signal update source & sink data */
-			update_buffers = 1;
-		}
-		/* buffer source data internally for future use by clients */
-		if (source->avail <= KPB_MAX_BUFFER_SIZE) {
-			/* TODO: should we copy what is available or just
-			 * a small portion of it?
-			 */
-			kpb_buffer_data(kpb, source, kpb->sink_period_bytes);
-		}
+	sink = (kpb->state == KPB_BUFFERING) ? kpb->rt_sink : kpb->cli_sink;
+
+	/* Process source data */
+	/* Check if there are valid pointers */
+	if (!source || !sink)
+		return -EIO;
+	if (!source->r_ptr || !sink->w_ptr) {
+		return -EINVAL;
+	} else if (sink->free == 0) {
```

Comment:
This ifs are not needed:
`if (sink->free == 0)`
`return -EIO;`
``
`if (source->avail == 0)`
`return -EIO;`
``
`copy_bytes = ...`